[expo-notifications] Add private InstallationIdProvider

sjchmiela opened the pull request:
# Why

To be able to register at Expo server we need to identify the device with a UUID.

# How

Implemented a non-exported module which:
- uses the same storage as kernel does for the UUID (in native, on web uses `localStorage`) 
- if the ID is missing, it generates and saves a new one.

# Test Plan

None yet, implementing `getExpoPushTokenAsync` will be a test for this module.

ide reviewed (changes requested):
We didn't add the legacy "installation ID" to expo-application because it is inherently flawed. This should be a private API that is used only by expo-notifications, except for legacy code that uses `Constants.installationId`. If applications would like to implement this behavior, a simple way is to generate a random UUID in JS and store it in SQLite.

For expo-notifications we need to not break backwards compatibility (need to use the UUID for now) and also eventually migrate new apps to use an identifier that changes when a user gets a new phone and restores it from backup. (This is not as trivial as just using `identifierForVendor`.)

sjchmiela commented on packages/expo-application/ios/EXApplication/EXApplication.m:
```diff
@@ -3,6 +3,8 @@
 #import <EXApplication/EXApplication.h>
 #import <UIKit/UIKit.h>
 
+NSString * const kEXApplicationInstallationIdKey = @"kEXApplicationInstallationIdKey";
```
Thanks to `EXScopedApplication` and `ScopedApplicationModule` this doesn't matter in ExpoKit applications — `Kernel`/`ExponentSharedPreferences` installationID is used in `expo-application` there. This key will only be used in new `expo-application` applications. Or maybe your concern applies to applications ejected to bare (that we would like to use the same UUID there)?

ide replied:
My concerns are that existing standalone apps need to read the existing data (we cannot generate a new UUID) and we should deprecate Constants.installationId and not provide a replacement (if developers want to re-implement this in user space they can do that, but we don't want to encourage it).

sjchmiela replied:
But we still need to have a UUID to be able to register the push token in the Expo server, right? Or would you want to deprecate this argument and identify devices by tokens?

If I understood the matter correctly the PR at current state should be ok. 🙂

ide replied:
> But we still need to have a UUID to be able to register the push token in the Expo server, right?

We need to have a unique ID but it does not necessarily need to be a UUID. 

>  Or would you want to deprecate this argument and identify devices by tokens?

We cannot identify devices by push tokens for two reasons. Push tokens can change at any time. We also don't want to store the previous push token because if a user gets a new phone and restores it from backup, we'll see two devices that report having had the same previous push token.

sjchmiela replied:
> We need to have a unique ID but it does not necessarily need to be a UUID.

At the moment server doesn't accept anything that isn't a UUIDv4, I experimented with using `Secure.ANDROID_ID` before and it failed (see [universe schema](https://github.com/expo/universe/blob/b60d7d4734940bfb234a9c545234a87aa2f3d8a7/server/www/src/apiv2/PushRouter.js#L31-L33)).

> Push tokens can change at any time.

That's why I would have imagined invalidating tokens that return `unregistered` from Push Notification servers in our storage.

> if a user gets a new phone and restores it from backup, we'll see two devices that report having had the same previous push token

If we identified devices by tokens (there wouldn't be a notion of `device` on server) in this situation there would be two tokens (after all a token must identify a single device, doesn't it?)

---

All in all I think making any of these changes would be non-trivial and if we ever wanted to revolutionize the API this would require giving it a good thinking. (+ my comments run around an idea of _removing UUID_ and it may not be a good idea – I was just curious about whether it would be doable.)

ide replied:
> That's why I would have imagined invalidating tokens that return unregistered from Push Notification servers in our storage.

We can only invalidate a token when a push notification is sent. If the developer doesn't send push notifications but the app changes its token, we don't get notified that the old token is invalid.

> If we identified devices by tokens (there wouldn't be a notion of device on server) in this situation there would be two tokens (after all a token must identify a single device, doesn't it?)

Each device would have a different new token and have the same old token. The server would need to handle this "fork".

sjchmiela committed:
[expo-notifications][ios] Add "private" getInstallationIdAsync method

sjchmiela committed:
[expo-notifications][android] Add "private" getInstallationIdAsync

Shared preferences filename and key match those used in ExponentSharedPreferences, so we'll use ExpoKit's values if they're present

sjchmiela committed:
[expo-notifications] Implement InstallationIdProvider in TS

ide commented on packages/expo-notifications/ios/EXNotifications/EXInstallationIdProvider.m:
```diff
@@ -0,0 +1,27 @@
+  if (!uuid) {
+    uuid = [[NSUUID UUID] UUIDString];
+    [[NSUserDefaults standardUserDefaults] setObject:uuid forKey:kEXDeviceInstallUUIDKey];
+    [[NSUserDefaults standardUserDefaults] synchronize];
```
```suggestion
```
https://developer.apple.com/documentation/foundation/nsuserdefaults/1414005-synchronize?language=objc

ide commented on packages/expo-notifications/src/InstallationIdProvider.web.ts:
```diff
@@ -0,0 +1,21 @@
+
+export default {
+  getInstallationIdAsync: async () => {
+    let installationId = localStorage.getItem(INSTALLATION_ID_KEY);
```
localStorage can fail for several reasons. expo-constants does this:

```
    let installationId;
    try {
      installationId = localStorage.getItem(ID_KEY);
      if (installationId == null || typeof installationId !== 'string') {
        installationId = UUID.create().toString();
        localStorage.setItem(ID_KEY, installationId as string);
      }
    } catch (error) {
      installationId = _sessionId;  // <----- this is just a UUID that is generated for the duration of the page's lifetime
    }
    return installationId;
```

sjchmiela replied:
Went with this ⬆️ + local _session ID_.

https://github.com/expo/expo/blob/632138f5cbf3ddbbbdb609b2b1e260f61b5161e4/packages/expo-notifications/src/InstallationIdProvider.web.ts#L7-L28

I'm torn between using a lazy getter for this fallback installation ID and just calculating and assigning it on every run.

ide commented on packages/expo-notifications/src/InstallationIdProvider.web.ts:
```diff
@@ -0,0 +1,21 @@
+  },
+  // mock implementations
+  addListener: () => null,
+  removeListeners: () => null,
```
```suggestion
  removeListeners: () => {},
```

ide commented on packages/expo-notifications/src/InstallationIdProvider.web.ts:
```diff
@@ -0,0 +1,21 @@
+import UUID from 'uuid-js';
+
+import { InstallationIdProvider } from './InstallationIdProvider';
```
Put shared types in another module so we don't need to import NativeModulesProxy on web.

ide commented on packages/expo-notifications/src/InstallationIdProvider.web.ts:
```diff
@@ -0,0 +1,21 @@
+
+import { InstallationIdProvider } from './InstallationIdProvider';
+
+const INSTALLATION_ID_KEY = '@@expo-notifications.InstallationId@@';
```
@EvanBacon We need to have a naming convention for Expo-internal keys in shared namespaces (like localStorage), what do you propose?

sjchmiela replied:
Sorry for using this as the first proposal, I had `nativeEmitterSubscriptionKey` of `@unimodules/core/EventEmitter` in memory at that time which is [`@@nativeEmitterSubscription@@`](https://github.com/expo/expo/blob/3e7135a3d61f87e6a6cf6be7c258069633a8c104/packages/@unimodules/react-native-adapter/src/EventEmitter.ts#L4).

I went through `localStorage` usages in `expo/packages` and they are:
- `EXPO_CONSTANTS_INSTALLATION_ID` for `expo-constants.installationId`
- `EXPO_ERROR_RECOVERY_STORAGE` for `expo-error-recovery.recoveryProps`

so I'll go with

- `EXPO_NOTIFICATIONS_INSTALLATION_ID` for `expo-notifications.installationId`

🙂

ide reviewed (changes requested):
The IDs look correctly backwards-compatible to me now.
